lib/recording/RecordingManager.js

```diff
@@ -177,18 +187,57 @@ class RecordingManager {
 
         if (!recordingEntrySession) {
             logger.info(`[${entryId}] Entry does not exist in recording list. Maybe entry has passed the limit of max duration`);
-            return
+            return Q.resolve();
         }
+        var addChunksPromise = null;
         let durationMSec =  recordingEntrySession.getDurationMSec();
-        if ( durationMSec >= this.recordingMaxDurationInMs && !recordingEntrySession.disableRecording) {
-            logger.info(`recording reached max duration [${mathUtils.durationToString(durationMSec, 2, mathUtils.HOUR_IN_MSEC)}] hrs`);
-            this.splitRecording(entryId, 'recording  reached max duration');
+        let stopRecording = false;
+        if (!recordingEntrySession.mp4FilesQueueLock &&
+            recordingEntrySession.maxDurationReached() && !recordingEntrySession.disableRecording) {
+            logger.info(`[${entryId}] recording reached max duration [${mathUtils.durationToString(durationMSec, 2, mathUtils.HOUR_IN_MSEC)}] hrs`);
+            addChunksPromise = recordingEntrySession.isUnlimitedRecordingDurationEnabled()
+                .then((isEnabled)=> {
+                    if (isEnabled) {
+                        return this.splitRecording(entryId, 'recording reached max duration');
+                    }
+                    else {
+                        stopRecording = true;
+                        return this.stopRecording(entryId);
+                    }
+            })
+        }
+        else {
+            addChunksPromise = Q.resolve();
         }
-        // update lastUpdateTime
-        recordingEntrySession.lastUpdateTime = new Date();
 
-        let Mp4FilesClone = _.clone(Mp4Files);
-        recordingEntrySession.updateRecordingFiles(Mp4FilesClone, flavorId);
+        return addChunksPromise
```

Comment:
I think it's more accurate to define it as deferred. and the return addChunksPromise.promise.
You can see the BackendClient for example, we use it all the time there